packages/flutter/lib/src/widgets/visibility.dart

```diff
@@ -0,0 +1,209 @@
+///  * [ExcludeSemantics], which can hide the child from accessibility tools.
+///  * [IgnorePointer], which can disable touch interactions with the child.
+///
+/// Using this widget is not necessary to hide children. The simplest way to
+/// hide a child is just to not include it, or, if a child _must_ be given (e.g.
+/// because this is a [StatelessWidget]) then to use [SizedBox.shrink] instead
+/// of the child that would otherwise be included.
+///
+/// See also:
+///
+///  * [AnimatedSwitcher], which can fade from one child to the next as the
+///    subtree changes.
+///  * [AnimatedCrossFade], which can fade between two specific children.
+class Visibility extends StatelessWidget {
+  /// Control whether the given [child] is [visible].
+  ///
+  /// The [child] and [replacement] arguments must not be null.
+  ///
+  /// The boolean arguments must not be null.
+  ///
+  /// The [maintainSemantics] and [maintainInteractivity] arguments can only be
+  /// set if [maintainSize] is set.
+  ///
+  /// The [maintainSize] argument can only be set if [maintainAnimation] is set.
+  ///
+  /// The [maintainAnimation] argument can only be set if [maintainState] is set.
+  const Visibility({
+    Key key,
+    @required this.child,
+    this.replacement = const SizedBox.shrink(),
+    this.visible = true,
+    this.maintainState = false,
+    this.maintainAnimation = false,
+    this.maintainSize = false,
+    this.maintainSemantics = false,
+    this.maintainInteractivity = false,
+  }) : assert(child != null),
+       assert(replacement != null),
+       assert(visible != null),
+       assert(maintainState != null),
+       assert(maintainAnimation != null),
+       assert(maintainSize != null),
+       assert(maintainState == true || maintainAnimation == false, 'Cannot maintain animations if the state is not also maintained.'),
+       assert(maintainAnimation == true || maintainSize == false, 'Cannot maintain size if animations are not maintained.'),
+       assert(maintainSize == true || maintainSemantics == false, 'Cannot maintain semantics if size is not maintained.'),
+       assert(maintainSize == true || maintainInteractivity == false, 'Cannot maintain interactivity if size is not maintained.'),
+       super(key: key);
+
+  /// The widget to show or hide, as controlled by [visible].
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The widget to use when the child is not [visible], assuming that none of
+  /// the `maintain` flags (in particular, [maintainState]) are set.
+  ///
+  /// The normal behavior is to replace the widget with a zero by zero box
+  /// ([SizedBox.shrink]).
+  ///
+  /// See also:
+  ///
+  ///  * [AnimatedCrossFade], which can animate between two children.
+  final Widget replacement;
+
+  /// Switches between showing the [child] or hiding it.
+  ///
+  /// The `maintain` flags should be set to the same values regardless of the
```

Comment:
This part seems really important. Perhaps copy paste into the class's dart doc too so people exploring what this widget does can know right away what it isn't for.

Also, is it worth creating a StatelessElement.update override to debug assert that these indeed don't change (or show a warning at least)?

Reply:
Done and done.

I don't see why we'd assert that they can't change. It's a pretty common pattern to have flags like this which, if changed, will cause the subtree to be discarded. At the most trivial, `key` is one such property.

Reply:
I think the difference in this case is that it's a side effect in other cases but is a breakage in the semantically stated goal (of maintainingState) in this case. It would be debug-helpful to not have it fail silently.

Also, not so much a feedback but just curious for setting precedents, do we have an existing pattern for cases where changing a parameter on ephemeral widgets between frames causes misbehaviors in that widget's stated goal?

As a developer, I think I'd have that preference for that `key` example too. If I had an implicitly animating widget and I change the key and broke the animation, it would be nice to at least get a message at debug time if not an assert to say that changing key on the same widget type at the same spot in the tree is really weird, are you sure you want to do it?

Reply:
Changing the key isn't really weird, it's literally how you indicate that you want to throw away the state.

Another example would be the widgets you build in `MaterialApp.builder`, if you change some of the flags on `MaterialApp` that change the tree shape. Or pretty much any of the flags on `Container`.

Reply:
`<feel free to ignore and submit. This is more educational for me than anything>`
* For the MaterialApp and Container etc, there's no conflict with the user's intent. There are side effects but you still correctly get a 'Container' when you change how the container draws itself and affects its child. Here, we're potentially breaking the stated intent of the user (of maintaining state). 

* I'm not sure changing key = losing state is super obvious for the layperson. We describe how to preserve state via keys in https://docs.flutter.io/flutter/widgets/StatefulWidget-class.html but maybe we should point it out more explicitly that changing key implies losing state and/or describe in https://docs.flutter.io/flutter/widgets/Widget/key.html that removing elements from the tree also implies losing state. 
`</>`